[docs] GCP MySQL IAM authentication support (#38943)
gravitational/teleport: docs/pages/database-access/guides/mysql-cloudsql.mdx

  Service
- (!docs/pages/includes/tctl.mdx!)

## Step 1/7. Enable Cloud SQL IAM authentication

Teleport uses [IAM
authentication](https://cloud.google.com/sql/docs/mysql/iam-authentication)
with Cloud SQL MySQL instances.

(!docs/pages/includes/database-access/cloudsql_enable_iam_auth.mdx type="MySQL" !)

> ## Step 2/7. Create a service account to access the MySQL instance

Reviewer: ```suggestion
## Step 2/7. Create a service account for a database user
```

nit: I think it just makes it a little more clear that one svc acct is for db user, one is for teleport db service

  Service
- (!docs/pages/includes/tctl.mdx!)

## Step 1/7. Enable Cloud SQL IAM authentication

Teleport can access MySQL on GCP Cloud SQL as built-in database users with
passwords or database users with [IAM
authentication](https://cloud.google.com/sql/docs/mysql/iam-authentication)
enabled.

> Teleport generates a one-time password for built-in database users and updates

Reviewer 2: Can we just remove the mention of the OTP from everywhere in the docs?

Author: removed.

### Grant permissions

Assign the Service Account the "Cloud SQL Admin" role:

![Grant Cloud SQL Admin to Service Account](../../../img/database-access/guides/cloudsql/service-account-sql-admin-grant@2x.png)
(!docs/pages/includes/database-access/cloudsql_grant_db_service_account.mdx!)

<Admonition type="note" title="Service account permissions">
  The default "Cloud SQL Admin" IAM role includes more permissions than the
  Database Service needs to generate one-time user passwords. To further restrict
  the service account, you can create a role that includes only the following
  permissions:
  "Service Account Token Creator", "Cloud SQL Viewer", and "Cloud SQL Admin"
  IAM roles include more permissions than the Database Service needs. To further
  restrict the service account, you can create a role that includes only the
  following permissions:
  ```ini
  # Used to download a list of database users.

>   cloudsql.users.list

Reviewer: Was this permission only used for password auth, or was it just not necessary at all? I tried to find where we were using it before, but I couldn't find it.

Author: I couldn't find anywhere using it either. @r0mant do you recall where `cloudsql.users.list` is used?

Reviewer 2: Looking at the guide history, it's been there since the very beginning but I can't remember why I put it there. If everything works without it, it's probably not needed.

Author: Sounds good. It does work without it when I was testing it.

![Instance Public IP](../../../img/database-access/guides/cloudsql/instance-public-ip@2x.png)

- (Optional) Cloud SQL instance root certificate.

(!docs/pages/includes/database-access/cloudsql_download_root_ca.mdx!)

## Step 5/7. Create a Teleport user

(!docs/pages/includes/database-access/create-user.mdx!)

> ## Step 6/7. Set up the Teleport Database service

Reviewer 3: ```suggestion
## Step 6/7. Set up the Teleport Database Service
```
"Database Service" is what we use in the Core Concepts page/throughout the docs

$ tsh db logout
```

## Troubleshooting

### Error when connecting to a replica instance

You may encounter the following error when connecting to a replica instance:

```code
$ tsh db connect --db-user root -n test cloudsql-replica 
ERROR 1105 (HY000): Could not update Cloud SQL user "<username>" password:

  The requested operation is not valid for a replica instance.

...
```

> Connecting as built-in database users with passwords are not supported for

Reviewer 2: Similar to above, I think we just shouldn't say it's supported at all. Only keep it for backwards-compatibility reasons but out of user docs.
